### PR TITLE
✨ add highlights api endpoint

### PR DESCRIPTION
**Pull request by webdevelukas**
Solves #60

[x] Add new `api/highlights` endpoint
[x] Create collection, schema and indexes in mongodb
[x] Highlights are now automaticly pushed into highlights collection
[ ] Make sure all components work correct

We should delete all documents in the mongo db

### CODE REVIEW

File: src/app/utils/api/highlights.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  page: number;
+  itemsPerPage: number;
+}): Promise<PaginatedMatchHighlightsClient> {
+  const { page, itemsPerPage, ...rest } = query;
+  const searchParams = new URLSearchParams({
+    page: page.toString(),
+    itemsPerPage: itemsPerPage.toString(),
+    ...rest,
+  });
+  const queryString = searchParams.toString();
```
Comment (AlexisRoe) [inline src/app/utils/api/highlights.ts:35]: looks good, is it possible to outsource it to another function, which returns the string directly, you use the same logic in two different endpoints

File: src/app/utils/highlights.ts [hunk 2]
```diff
@@ -112,14 +112,15 @@ export function startCaptureHighlights(): void {
...
         events: events,
         timestamp: timestamp,
       };
-      const updatedMatch = await postMatchHighlight(highlight, activeMatch._id);
 
-      activeMatch = updatedMatch;
-      console.log('updatedMatch', updatedMatch);
+      const createdHighlight = await postHighlight(highlight);
+
+      console.log('createdHighlight', createdHighlight);
```
Comment (AlexisRoe) [inline src/app/utils/highlights.ts]: Is the console.log really necessary, you do nothing with it, except logging it to the console, wouldn`t it be enough to just post it?
```suggestion
      await postHighlight(highlight);
```

File: src/lib/router.ts [hunk 3]
```diff
@@ -129,4 +85,40 @@ router.get('/matches', async (request, response, next) => {
...
+      typeof videoSrc !== 'string' ||
+      typeof matchId !== 'string';
+
+    if (requestIsInvalid) {
+      response.status(400).send('Invalid payload');
+      return;
+    }
+
+    const highlight: Omit<MatchHighlight, '_id'> = {
+      matchId: new ObjectId(matchId),
```
Comment (AlexisRoe) [inline src/lib/router.ts]: there is no validation if the matchid is a valid ObjectId, which could resolve in an exception when inserted into the query
```suggestion
    const { timestamp, events, videoSrc, matchId } = request.body;

    const requestIsInvalid =
      typeof timestamp !== 'number' ||
      !Array.isArray(events) ||
      !events.every((event) => typeof event === 'string') ||
      typeof videoSrc !== 'string' ||
      typeof matchId !== 'string';
      
     if(!ObjectId.isvalid(matchId)) {
       response.status(400).send('Invalid matchId');
       return;
     }

    if (requestIsInvalid) {
      response.status(400).send('Invalid payload');
      return;
    }

    const highlight: Omit<MatchHighlight, '_id'> = {
      matchId: new ObjectId(matchId),
```

**Review by AlexisRoe — CHANGES_REQUESTED** [review]
looks great

File: src/lib/matches.ts [hunk 4]
```diff
@@ -41,17 +41,8 @@ export function ensureMatchesSchema(): Promise<Document> {
...
-                },
-                events: {
-                  bsonType: 'array',
-                  items: {
-                    bsonType: 'string',
-                  },
-                },
-                videoSrc: {
-                  bsonType: 'string',
+                highlightId: {
```
Comment (lmachens) [inline src/lib/matches.ts]: `highlights`  is not part of `matches` anymore and should be removed

File: src/lib/highlights.ts [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+          },
+          videoSrc: {
+            bsonType: 'string',
+          },
+          createdAt: {
+            bsonType: 'date',
+          },
+        },
+        additionalProperties: false,
+        required: ['matchId', 'timestamp', 'events', 'videoSrc'],
```
Comment (lmachens) [inline src/lib/highlights.ts]: Maybe add `createdAt`  as required too

File: src/app/utils/api/matches.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+  );
+  matches.results = matches.results.map((match) => ({
+    ...match,
+    createdAt: new Date(match.createdAt),
+  }));
+  return matches;
+}
+
+export async function postMatch(gameId: number): Promise<MatchClient> {
+  const { username } = (await getCurrentUser()) || { username: 'unknown' };
```
Comment (lmachens) [inline src/app/utils/api/matches.ts:28]: I think the a match requires the `avatarSrc` too. Not important to add it to this pull request, just noticed.

**Review by lmachens — CHANGES_REQUESTED** [review]
Good job!
I reviewed server side

**Review by lmachens — APPROVED** [review]
gj

File: src/app/utils/api/highlights.ts (hunk 1 above)
Comment (webdevelukas) [reply src/app/utils/api/highlights.ts:35]: Good point, I will make this change later :-)

File: src/app/components/Feed/Feed.tsx [hunk 7]
```diff
@@ -48,15 +49,17 @@ function Feed({ username, onMatchClick }: FeedProps): JSX.Element {
...
+          key={highlight._id}
+          highlight={highlight}
+          onHighlightClick={onHighlightClick}
+          matchIsActive={true}
+          layout="full"
         />
       ))}
-      {matches?.results.length === 0 && (
-        <p className={classes.noMatches}>No matches found</p>
+      {highlights?.results.length === 0 && (
```
Comment (AlexisRoe) [inline src/app/components/Feed/Feed.tsx:61]: just a minor improvement without any value
```suggestion
      {!highlights?.results.length && (
```

**Review by AlexisRoe — APPROVED** [review]
thanks for the work, looks great so far

**Pull request merged**